peer/brontide.go

```diff
@@ -2252,60 +2396,69 @@ func (p *Brontide) queueHandler() {
 // remote peer in order to keep the connection alive and/or determine if the
 // connection is still active.
 //
+// NOTE: timeout MUST be shorter than interval
+//
 // NOTE: This method MUST be run as a goroutine.
-func (p *Brontide) pingHandler() {
+func (p *Brontide) pingHandler(newPayload func() []byte, pongSize func() uint16,
+	interval time.Duration, timeout time.Duration) {
+
 	defer p.wg.Done()
 
-	pingTicker := time.NewTicker(pingInterval)
+	pingTicker := time.NewTicker(interval)
 	defer pingTicker.Stop()
 
-	// TODO(roasbeef): make dynamic in order to create fake cover traffic
-	const numPongBytes = 16
-
-	blockEpochs, err := p.cfg.ChainNotifier.RegisterBlockEpochNtfn(nil)
-	if err != nil {
-		p.log.Errorf("unable to establish block epoch "+
-			"subscription: %v", err)
-		return
-	}
-	defer blockEpochs.Cancel()
-
-	var (
-		pingPayload [wire.MaxBlockHeaderPayload]byte
-		blockHeader *wire.BlockHeader
-	)
 out:
 	for {
 		select {
-		// Each time a new block comes in, we'll copy the raw header
-		// contents over to our ping payload declared above. Over time,
-		// we'll use this to disseminate the latest block header
-		// between all our peers, which can later be used to
-		// cross-check our own view of the network to mitigate various
-		// types of eclipse attacks.
-		case epoch, ok := <-blockEpochs.Epochs:
-			if !ok {
-				p.log.Debugf("block notifications " +
-					"canceled")
-				return
-			}
-
-			blockHeader = epoch.BlockHeader
-			headerBuf := bytes.NewBuffer(pingPayload[0:0])
-			err := blockHeader.Serialize(headerBuf)
-			if err != nil {
-				p.log.Errorf("unable to encode header: %v",
-					err)
-			}
-
 		case <-pingTicker.C:
 
+			pongBytes := pongSize()
 			pingMsg := &lnwire.Ping{
-				NumPongBytes: numPongBytes,
-				PaddingBytes: pingPayload[:],
+				// TODO(roasbeef): make dynamic in order to
+				// create fake cover traffic
+				// NOTE(proofofkeags): this was changed to be
+				// dynamic to allow better pong identification,
+				// however, more thought is needed to make this
+				// actually usable as a traffic decoy
+				NumPongBytes: pongBytes,
+				PaddingBytes: newPayload(),
 			}
 
-			p.queueMsg(pingMsg, nil)
+			msgWritten := make(chan error)
+			p.queueMsg(pingMsg, msgWritten)
+			p.wg.Add(1)
+			go func() {
+				defer p.wg.Done()
+				err, ok := <-msgWritten
+				if err != nil {
+					p.log.Errorf(
+						"writeHandler failed to write "+
+							"ping to the wire: %v",
+						err,
+					)
+
+					return
+				}
+				if !ok {
+					p.log.Error("writeHandler prematurely" +
+						" closed confirmation channel")
+
+					return
+				}
+
+				onTimeout := func() {
+					eStr := "pong timeout for %s " +
+						"-- disconnecting"
+					p.log.Warnf(eStr, p)
+					p.Disconnect(
+						fmt.Errorf(eStr, p),
+					)
+				}
+				p.pingManager.markPingSend(
```

Comment:
There is a race condition here.  A later ping could actually execute `markPingSend` before an earlier ping, since the call is made in a separate goroutine.  The odds of the race occurring in practice seem low, but the odds go up if the peer doesn't ack the first ping before the next one is queued (which would cause the first goroutine to block on `<-msgWritten`).